File: TableViewCache/TableViewCache/ViewController.m
//
//  ViewController.m
//  TableViewCache
//
//  Created by lin kang on 19/12/18.
//  Copyright © 2018 lin kang. All rights reserved.
//

#import "ViewController.h"
#import <ZhuoZhuo/ZhuoZhuo.h>
#import "SKTableViewCell.h"
#import "DetailViewController.h"

@interface ViewController ()<UITableViewDelegate,UITableViewDataSource,SKTableViewCellDelegate>
{
    UITableView *_tableView;
    NSMutableArray *_dataSource;
    NSMutableArray *hightArray;
}
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    if (_tableView == nil) {
        _tableView = [UITableView new];
        _tableView.estimatedRowHeight = 0;
        _tableView.estimatedSectionHeaderHeight = 0;
        _tableView.estimatedSectionFooterHeight = 0;
        _tableView.delegate = self;
        _tableView.dataSource = self;
    }
    _tableView.frame = self.view.bounds;
    [self.view addSubview:_tableView];
    
    if (_dataSource == nil) {
        _dataSource = [NSMutableArray array];
    }
    //数据源
    NSMutableArray *tempDataSourceArray = [RdTestGetResource__NotAllowedInMainThread() copy];
    [_dataSource addObjectsFromArray:tempDataSourceArray];
    [_dataSource addObjectsFromArray:tempDataSourceArray];
    [_dataSource addObjectsFromArray:tempDataSourceArray];

    //计算cell高度
    if (hightArray == nil) {
        hightArray = [NSMutableArray array];
    }
    
    NSMutableArray *tempHihgtArray = [NSMutableArray array];
    for (ResponseModel *model  in _dataSource){
        CGFloat height = model.ImageSizeInfo.Height + 70 ;
        UITextView *detailTextView = [UITextView new];
        detailTextView.frame = CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width  - 30, 20);
        detailTextView.font = [UIFont systemFontOfSize:14];
        detailTextView.text = model.Context;
        [detailTextView sizeToFit];
        height += detailTextView.bounds.size.height ;
        [tempHihgtArray addObject:@(height)];
    }
    [hightArray addObjectsFromArray:tempHihgtArray];
    [hightArray addObjectsFromArray:tempHihgtArray];
    [hightArray addObjectsFromArray:tempHihgtArray];

}



- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    
    return _dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    static NSString *cellID = @"myCell";
    SKTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellID];
    if (cell == nil) {
        cell = [[SKTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellID];
        cell.delegate = self;
    }
    cell.tag = indexPath.row;
    [cell setCell:[_dataSource objectAtIndex:indexPath.row]];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    CGFloat height = [[hightArray objectAtIndex:indexPath.row] floatValue];
    return height;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    
}

-(void)didTapTextAtIndex:(NSInteger)index forRow:(NSInteger)row{
    NSLog(@"indexTouched:%ld,atRow:%ld",(long)index,(long)row);
    ResponseModel *model = [_dataSource objectAtIndex:row];
    ClickInfo *clickInfo = [model.ClickInfoList objectAtIndex:index];
    DetailViewController *detailVC = [DetailViewController new];
    detailVC.detailURL = [NSURL URLWithString:clickInfo.Url];
    [self.navigationController pushViewController:detailVC animated:true];
}
@end

File: Animations/Animations/ViewController.swift
//
//  ViewController.swift
//  Animations
//
//  Created by kang lin on 2018/1/16.
//  Copyright © 2018年 康林. All rights reserved.
//

import UIKit
struct  AnimationType: OptionSet {
    var rawValue: Int
    
    init(rawValue:Int) {
        self.rawValue = rawValue;
    }
    
    static let  PalseAnimation = AnimationType(rawValue: 1)
    
    func getDesc() -> String {
        switch self {
        case AnimationType.PalseAnimation:
            return "脉冲动画"
        default:
            return ""
        }
    }
}

class ViewController: UIViewController {

    let tableView:UITableView = UITableView()
    let animationTypes = [AnimationType.PalseAnimation];
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        tableView.delegate = self;
        tableView.dataSource = self;
        tableView.frame = self.view.bounds
        self.view.addSubview(tableView)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

extension ViewController:UITableViewDelegate,UITableViewDataSource{
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
     {
        return animationTypes.count
    }
    
    // Row display. Implementers should *always* try to reuse cells by setting each cell's reuseIdentifier and querying for available reusable cells with dequeueReusableCellWithIdentifier:
    // Cell gets various attributes set automatically based on table (separators) and data source (accessory views, editing controls)
    
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
     {
        let cellID = "haha"
        var cell = tableView.dequeueReusableCell(withIdentifier: cellID)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: cellID)
            cell?.accessoryType = .disclosureIndicator
        }
        cell?.textLabel?.text = animationTypes[indexPath.row].getDesc()
        return cell!;
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let type =  animationTypes[indexPath.row]
        switch type {
        case AnimationType.PalseAnimation:
            let vc = YKPulseViewController()
            self.navigationController?.pushViewController(vc, animated: true);
            break
        default: break
            
        }
    }
}
